tally/src/buckets.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "tally/buckets.h"
+
+#include <stdexcept>
+
+#include "tally/buckets_iterator.h"
+#include "tally/src/buckets_calculator.h"
+
+namespace tally {
+
+Buckets::Buckets(Buckets::Kind kind, BucketsCalculator calculator, uint64_t num)
+    : kind_(kind), calculator_(calculator), num_(num) {
+  if (num == 0) {
+    throw std::invalid_argument("Number of buckets cannot be zero");
```

Comment:
could you confirm stance on exceptions w/ Anthony & co

Reply:
Yep, I've already spoken with him and he and a few other members of his team are going to review as well.

Reply:
cool, maybe keep a running list of open questions you have for them in the PR notes so they don't have to dig thru all comments for it.

Reply:
exceptions are fine for exceptional cases; we just don't use them for flow control. :)

Reply:
So what I've found with Jaeger (and it is part of the OpenTracing standard as well) is that people really hate to see instrumentation throwing exceptions. If you can, make all public facing methods `noexcept`. That way your metrics code won't ever bring down a running application because someone forgot a `try`/`catch`.

Reply:
That's a fair point, I'll make all public methods which shouldn't ever throw an exception `noexcept`. I do think there are cases, such as this one, where the user provides an invalid argument, where exceptions make sense though.

Reply:
callsite exceptions (e.g., validation) are absolutely an acceptable use of exceptions.

most of this will probably do with strings; empty maps/default integral values/etc are all normal. however, if i have a scope with an empty namespace and attempt to use a metric with an empty name, what should happen?

Reply:
Presently that is permitted by the other tally clients and the same is true here, a metric can have an empty name.